cpp/src/arrow/compute/exec/spilling_util.cc

```diff
@@ -0,0 +1,483 @@
+        std::vector<ArrayInfo> arrays;
+        std::vector<struct iovec> ios;
+    };
+
+
+Result<FileHandle> OpenTemporaryFile()
+{
+    static std::once_flag generate_tmp_file_name_flag;
+
+    constexpr int kFileNameSize = 1024;
+    static char name_template[kFileNameSize];
+    char name[kFileNameSize];
+
+    char *name_template_ptr = name_template;
+    std::call_once(generate_tmp_file_name_flag, [name_template_ptr]() noexcept
+    {
+        const char *selectors[] = { "TMPDIR", "TMP", "TEMP", "TEMPDIR" };
+        constexpr size_t kNumSelectors = sizeof(selectors) / sizeof(selectors[0]);
+#ifdef __ANDROID__
+        const char *backup = "/data/local/tmp/";
+#else
+        const char *backup = "/var/tmp/";
+#endif
+        const char *tmp_dir = backup;
+        for(size_t i = 0; i < kNumSelectors; i++)
+        {
+            const char *env = getenv(selectors[i]);
+            if(env)
+            {
+                tmp_dir = env;
+                break;
+            }
+        }
+        size_t tmp_dir_length = std::strlen(tmp_dir);
+
+        const char *tmp_name_template = "/ARROW_TMP_XXXXXX";
+        size_t tmp_name_length = std::strlen(tmp_name_template);
+
+        if((tmp_dir_length + tmp_name_length) >= kFileNameSize)
+        {
+            tmp_dir = backup;
+            tmp_dir_length = std::strlen(backup);
+        }
+
+        std::strncpy(name_template_ptr, tmp_dir, kFileNameSize);
+        std::strncpy(name_template_ptr + tmp_dir_length, tmp_name_template, kFileNameSize - tmp_dir_length);
+    });
+
+    std::strncpy(name, name_template, kFileNameSize);
+
+#ifdef __APPLE__
+    int fd = mkstemp(name);
+    if(fd == kInvalidHandle)
+        return Status::IOError(strerror(errno));
+    if(fcntl(fd, F_NOCACHE, 1) == -1)
+        return Status::IOError(strerror(errno));
+#else
+    // If we failed, it's possible the temp directory didn't like O_DIRECT,
+    // so we try again without O_DIRECT, and if it still doesn't work then
+    // give up.
+    int fd = mkostemp(name, O_DIRECT);
+    if(fd == kInvalidHandle)
+    {
+        std::strncpy(name, name_template, kFileNameSize);
+        fd = mkstemp(name);
+        if(fd == kInvalidHandle)
+            return Status::IOError(strerror(errno));
+    }
+#endif
+
+    if(unlink(name) != 0)
+        return Status::IOError(strerror(errno));
+    return fd;
+}
+
+static Status CloseTemporaryFile(FileHandle handle)
+{
+    if(close(handle) == -1)
+        return Status::IOError(strerror(errno));
+    return Status::OK();
+}
+
+static Status WriteBatch_PlatformSpecific(FileHandle handle, SpillFile::BatchInfo &info)
+{
+    if(pwritev(handle, info.ios.data(), static_cast<int>(info.ios.size()), info.start) == -1)
+        return Status::IOError("Failed to spill!");
+
+    // Release all references to the buffers, freeing them. 
+    for(ArrayInfo &arr : info.arrays)
+        for(int i = 0; i < 3; i++)
+            if(arr.bufs[i])
+                arr.bufs[i].reset();
```

Comment:
It looks like we do not free buffers in Windows specific version.

Reply:
I have a line under the `WriteFile`, sayinmg `info.bufs[i].reset()`, though that's a typo. I changed it to `arr.bufs[i].reset()`. I'll do a pass fixing Windows compiler errors after I get Linux working.